### PR TITLE
test/zdtm: mount a new tmpfs into the zdtm /dev

### LINKED ISSUE #2441: Massive testsuite failures when testing from a subdirectory of /tmp on Fedora (starting with zdtm/static/aio00)
Fedora 39 mounts `/tmp` like this:

```
tmpfs on /tmp type tmpfs (rw,nosuid,nodev,seclabel,nr_inodes=1048576,inode64)
```

This seems to cause all tests to fail if the CRIU Git checkout is located in `/tmp`. It's a bit non-obvious, so maybe the testsuite could check the mount flags of the mount point? At the very least, this ticket should be easier to find due to the `zdtm/static/aio00` string …

### PR DESCRIPTION
**Pull request by avagin**
The current file system can be mounted with nodev.

Fixes #2441

### CODE REVIEW

File: test/zdtm.py [hunk 1]
```diff
@@ -256,11 +266,18 @@ def __construct_root(self):
         for ldir in ["/bin", "/sbin", "/lib", "/lib64"]:
             os.symlink(".." + ldir, self.root + "/usr" + ldir)
 
+    def __construct_dev(self):
+        for dir in self.__dev_dirs:
+            os.mkdir(os.path.join(self.devpath, dir))
+            os.chmod(os.path.join(self.devpath, dir), 0o777)
```
Comment (rst0git) [inline test/zdtm.py:272]: Could we use `os.mkdir(os.path.join(self.devpath, dir), mode=0o777)` here?
https://docs.python.org/3/library/os.html#os.mkdir

Comment (avagin) [reply test/zdtm.py:272]: In this script, we use chmod for all new directories. I think the problem here is that umask is masked out from mode specified for mkdir.

File: test/zdtm.py [hunk 2]
```diff
@@ -85,8 +87,12 @@ def make_tests_root():
         tests_root = (os.getpid(), tempfile.mkdtemp("", "criu-root-", tmpdir))
         atexit.register(clean_tests_root)
         os.mkdir(os.path.join(tests_root[1], "root"))
+        devpath = os.path.join(tests_root[1], "dev")
+        os.mkdir(devpath)
+        subprocess.check_call(["mount", "-t", "tmpfs", "criu-test-dev", devpath])
+        os.chmod(devpath, 0o777)
```
[the remark this answers is not in the document]
Comment (avagin) [reply test/zdtm.py]: Fixed. Thanks.

File: test/zdtm.py [hunk 3]
```diff
@@ -510,6 +531,7 @@
             # unprivileged users. The permissions here are set to test
             # this use case.
             os.chmod(os.path.dirname(self.__flavor.root), 0o700)
+            os.chmod(os.path.dirname(self.__flavor.devpath), 0o755)
```
[the remark this answers is not in the document]
Comment (Snorch) [reply test/zdtm.py:536]: make_tests_root already made it 0o755, do we need this line anyway?

File: test/zdtm/lib/ns.c [hunk 4]
```diff
@@ -51,6 +51,15 @@ static int prepare_mntns(void)
 		return -1;
 	}
 
+	dev_path = getenv("ZDTM_DEV");
+	if (dev_path) {
+		snprintf(path, sizeof(path), "%s/dev", root);
+		if (mount(dev_path, path, NULL, MS_BIND | MS_REC, NULL)) {
```
Comment (Snorch) [inline test/zdtm/lib/ns.c:57]: As we've just created ZDTM_DEV mount we know for sure that it does not have children mounts, maybe we can remove MS_REC here?

File: test/zdtm/lib/ns.c [hunk 5]
```diff
@@ -51,6 +51,15 @@ static int prepare_mntns(void)
...
+			pr_perror("Unable to mount %s",  path);
+			return -1;
+		}
```
Comment (Snorch) [inline test/zdtm/lib/ns.c:60]: Can we please make this external mount private explicitly.

File: test/zdtm.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (avagin) [reply test/zdtm.py:536]: If I remove this line, zdtm uns tests fail with this error:
```
========================= Run zdtm/static/env00 in uns =========================
Start test
./env00 --pidfile=env00.pid --outfile=env00.out --envname=ENV_00_TEST
Run criu dump
Run criu restore
=[log]=> dump/zdtm/static/env00/230/1/restore.log
------------------------ grep Error ------------------------
b'(00.010566)      1: No ipcns-sem-11.img image'
b'(00.017814)      1: net: Try to restore a link 10:1:lo'
b'(00.017826)      1: net: Restoring link lo type 1'
b'(00.021756)      1: net: \tRunning ip addr restore'
b'Error: ipv4: Address already assigned.'
b'Error: ipv6: address already assigned.'
b'(00.062690)      1: mnt-v2: Mount 865 is detected as dir-mount'
b'(00.062692)      1: mnt-v2: Create plain mountpoint /tmp/.criu.mntns.GxGH9W/mnt-0000000865 for 865'
b'(00.062699)      1: mnt-v2: \tMounting tmpfs @865 (0)'
b'(00.062702)      1: mnt-v2: \tBind /tmp/criu-root-1dncsez8/dev to /tmp/.criu.mntns.GxGH9W/mnt-0000000865'
b'(00.062712)      1: Error (criu/mount-v2.c:319): mnt-v2: Failed to open_tree /tmp/criu-root-1dncsez8/dev: Permission denied'
b'(00.064060) uns: calling exit_usernsd (-1, 1)'
b'(00.064129) uns: daemon calls 0x55f0764463b0 (255, -1, 1)'
b'(00.064155) uns: `- daemon exits w/ 0'
b'(00.065353) uns: daemon stopped'
b'(00.065371) Error (criu/cr-restore.c:2311): Restoring FAILED.'
------------------------ ERROR OVER ------------------------
################# Test zdtm/static/env00 FAIL at CRIU restore ##################
Test output: ================================
```

[the remark this answers is not in the document]
Comment (avagin) [reply test/zdtm.py:536]: Oops, it changes a parent directory of self.__flavor.devpath, which is actually the same directory that we changed one line above...

**Review by Snorch — APPROVED** [review]

**Pull request merged**